File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        n = len(s)
        dp = [-1] * (n + 1)
        word_set = set(dictionary)
        
        def solve(i: int) -> int:
            if i >= n:
                return 0
            
            if dp[i] != -1:
                return dp[i]
            
            result = 1 + solve(i + 1)
            
            for j in range(i, n):
                curr = s[i:j + 1]
                if curr in word_set:
                    result = min(result, solve(j + 1))
            
            dp[i] = result
            return result
        
        return solve(0)

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    vector<int> dp;
    
    int solve(int i, string &s, unordered_set<string> &st, int &n) {
        if (i >= n) {
            return 0;
        }
        
        if (dp[i] != -1) {
            return dp[i];
        }

        int result = 1 + solve(i + 1, s, st, n);

        for (int j = i; j < n; j++) {
            string curr = s.substr(i, j - i + 1);
            
            if (st.find(curr) != st.end()) {
                result = min(result, solve(j + 1, s, st, n));
            }
        }

        return dp[i] = result;
    }

    int minExtraChar(string s, vector<string>& dictionary) {
        int n = s.length();
        dp.resize(n+1, -1);
        unordered_set<string> st(dictionary.begin(), dictionary.end());
        return solve(0, s, st, n);
    }
};